TLS 1.3 client-side mutual authentication implementation

ryru opened the pull request:
This PR addresses:

- #30

Commit:
tls-crypto: Generalise private key search

Move this functionality from tls_server to tls_crypto to better reuse it in tls_peer.

Commit:
tls-peer: Derive app key after server FINISHED message

and set the new key for inbound traffic. Outbound traffic key is set
after client FINISHED message is sent to the server.

Commit:
tls-peer: Mutual authentication support for TLS 1.3

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -1397,12 +1397,27 @@ static status_t send_certificate(private_tls_peer_t *this,
-	this->private = find_private_key(this);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	if (version_max > TLS_1_1) {
```
```suggestion
	if (version_max > TLS_1_1)
	{
```

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -1397,12 +1397,27 @@ static status_t send_certificate(private_tls_peer_t *this,
 	{
-		DBG1(DBG_TLS, "no TLS peer certificate found for '%Y', "
-			 "skipping client authentication", this->peer);
+		DBG1(DBG_TLS, "no usable TLS peer key found for '%Y'", this->peer);
```
While we are searching for a key, using "certificate" in the log message might be more user accessible (but maybe use "client certificate" instead of "peer certificate").

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -1397,12 +1397,27 @@ static status_t send_certificate(private_tls_peer_t *this,
+	if (version_max > TLS_1_1) {
+		array_t *key_types;
+
+		enumerator = tls_get_supported_key_types(version_min, version_max);
+		key_types = tls_create_common_key_types(enumerator, this->hashsig);
+		enumerator->destroy(enumerator);
+		this->private = tls_find_private_key(this->peer, this->peer_auth,
+											 key_types, NULL);
```
As discussed, it might be possible to combine these into a single public helper method in `tls_crypto.c` (or perhaps two, as the server calls `tls_find_private_key()` multiple times with the array of key types, but even one should be possible if it's an enumerator that returns private keys and internally keeps track of the key types via array).

ryru replied:
I’m not quite sure how to keep track on the key types internally via an array. I think I have to do the while-loop on the array of `key_types` with `key = lib->credmgr->get_private()` directly in `tls_find_private_key()`. I could then iterate over all possible and found keys and store them in a new array. The function `tls_find_private_key()` could then return an enumerator via `array_create_enumerator()`. The caller then has to iterate over all keys but picks the first and removes it from the array with `array_remove_at()` in case it has to iterate over possible keys again later. Is this the correct approach?

tobiasbrunner replied:
You'd actually replace `tls_find_private_key()` and make that the main loop of the enumerator, which returns TRUE if another key type (from the internally stored array) and private key (via `get_private()`) was found and FALSE if there are no remaining key types. A caller can simply call `enumerate()` to retrieve another private key, if any.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -208,29 +173,23 @@ static bool select_suite_and_key(private_tls_server_t *this,
 	{
-		DBG1(DBG_TLS, "no usable TLS server certificate found for '%Y'",
-			 this->server);
+		DBG1(DBG_TLS, "no usable TLS server key found for '%Y'", this->server);
```
Same here regarding key -> certificate.

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -843,56 +844,28 @@ static status_t process_key_exchange(private_tls_peer_t *this,
 
-/**
- * Process a Certificate Request message
+/*
```
```suggestion
/**
```

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -843,56 +844,28 @@ static status_t process_key_exchange(private_tls_peer_t *this,
-/**
- * Process a Certificate Request message
+/*
+ * Read all available certificate authorities to output writer
```
"to output writer"? Maybe "from the given reader"?

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -912,6 +885,106 @@ static status_t process_certreq(private_tls_peer_t *this, bio_reader_t *reader)
+
+	if (this->tls->get_version_max(this->tls) < TLS_1_3)
+	{
+
```
```suggestion
```

Commit:
tls-peer: Use key enumeration also in TLS version lower 1.2

This, until now, worked by key selection based on
tls_client_certificate_type_t and now has a simple mapping from these
types to tls_signature_scheme_t types.

ryru commented on src/libtls/tls_peer.c:
```diff
@@ -1405,33 +1399,32 @@ static status_t send_certificate(private_tls_peer_t *this,
+		this->private = key->get_ref(key);
+		this->peer_auth->merge(this->peer_auth, auth, FALSE);
 	}
+	enumerator->destroy(enumerator);
```
Here an additional check is needed:

```
if (enumerator)
{
    enumerator->destroy(enumerator);
}
```

tobiasbrunner replied:
That is, just use `DESTROY_IF(enumerator);`.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2553,10 +2554,16 @@ CALLBACK(destroy_key_types, void,
+ * Create an enumerator over supported key types within a specific TLS range
+ *
+ * Enumerates over key_type_t
+ *
+ * @param min_version	minimum negotiated TLS version
+ * @param max_version	maximum negotiated TLS version
+ * @return				enumerator of key types
```
We don't use Doxygen comments for private helper functions.

```suggestion
```
The line above maybe too because the description already explains that key types are enumerated.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2582,3 +2589,115 @@ enumerator_t *tls_get_supported_key_types(tls_version_t min_version,
+
+/**
+ * Create an array of an intersection of server and peer supported key types
+ *
+ * @param enumerator	enumerator of key types
+ * @param hashsig		hash and signature algorithms supported by other peer
+ * @return				array of key types supported by both peers
```
```suggestion
```

tobiasbrunner commented on src/libtls/tls_crypto.h:
```diff
@@ -700,15 +700,18 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group);
- * Create an enumerator over supported key types within a specific TLS version range
- *
- * Enumerates over key_type_t
+ * Find a trusted private key to encrypt/verify key exchange data
```
Note that private keys are always trusted.

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -914,6 +887,105 @@ static status_t process_certreq(private_tls_peer_t *this, bio_reader_t *reader)
+			if (!extensions->read_uint16(extensions, &extension_type) ||
+				!extensions->read_data16(extensions, &extension_data))
+			{
+				DBG1(DBG_TLS, "invalid extension in ServerHello");
```
```suggestion
				DBG1(DBG_TLS, "invalid extension in CertificateRequest");
```

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -1426,37 +1504,45 @@ static status_t send_certificate(private_tls_peer_t *this,
-		if (cert)
+		if (cert->get_encoding(cert, CERT_ASN1_DER, &data))
+		{
+			DBG1(DBG_TLS, "sending TLS peer certificate '%Y'",
```
Maybe change this to
```suggestion
			DBG1(DBG_TLS, "sending TLS client certificate '%Y'",
```

tobiasbrunner reviewed (approved):
Only some minor stuff, looks good otherwise.
